Merge rewrite branch, splitting into two books (#414)
rust-embedded/discovery: microbit/src/explore.md

system with no OS, and thus no threads. There are two main approaches to
multitasking: preemptive multitasking and cooperative multitasking.

In preemptive multitasking a task that's currently being executed can, at any point in time, be
*preempted* (interrupted) by another task. On preemption, the first task will be suspended and the
processor will instead execute the second task. At some point the first task will be resumed.
Microcontrollers provide hardware support for preemption in the form of *interrupts*.

In cooperative multitasking a task that's being executed will run until it reaches a *suspension
point*. When the processor reaches that suspension point it will stop executing the current task and
instead go and execute a different task. At some point the first task will be resumed. The main
difference between these two approaches to multitasking is that in cooperative multitasking *yields*
execution control at *known* suspension points instead of being forcefully preempted at any point of
its execution.

### Sleeping

All our programs have been continuously polling peripherals to see if there's

> anything that needs to be done. However, some times there's nothing to be done!

Reviewer: ```suggestion
anything that needs to be done. However, sometimes there's nothing to be done!
```

At those times, the microcontroller should "sleep".

When the processor sleeps, it stops executing instructions and this saves power.
It's almost always a good idea to save power so your microcontroller should be
sleeping as much as possible. But, how does it know when it has to wake up to
perform some action? "Interrupts" (see below for what exactly those are)
are one of the events that wake up the microcontroller but there are others
and the `wfi` and `wfe` are the instructions that make the processor "sleep".

## Topics related to microcontroller capabilities

> Microcontrollers (like our nRF52/nRF51) have many different capabilities. However, many share similar

Reviewer: ```suggestion
Microcontrollers (like our nRF52/nRF51) have many capabilities. However, many share similar
```

In general, PWM can be used to control how much *power* is given to some
electric device. With proper (power) electronics between a microcontroller and
an electrical motor, PWM can be used to control how much power is given to the
motor thus it can be used to control its torque and speed. Then you can add an
angular position sensor and you got yourself a closed loop controller that can
control the position of the motor at different loads.

PWM is already abstracted within the [`embedded-hal` `Pwm` trait] and you will
again find implementations of this in the [`nrf52-hal`].

[`embedded-hal` `Pwm` trait]: https://docs.rs/embedded-hal/0.2.6/embedded_hal/trait.Pwm.html

### Digital inputs

We have used the microcontroller pins as digital outputs, to drive LEDs. But
these pins can also be configured as digital inputs. As digital inputs, these
pins can read the binary state of switches (on/off) or buttons (pressed/not
pressed).

Again digital inputs are abstracted within the [`embedded-hal` `InputPin` trait]
and of course the [`nrf52-hal`] does have an implementation for them.

(*spoilers* reading the binary state of switches / buttons is not as

> straightforward as it sounds ;-)

Reviewer: ```suggestion
straightforward as it sounds ;-) )
```

[`embedded-hal` `InputPin` trait]: https://docs.rs/embedded-hal/0.2.6/embedded_hal/digital/v2/trait.InputPin.html

### Analog-to-Digital Converters (ADC)

> There are a lots of digital sensors out there. You can use a protocol like I2C

Reviewer: ```suggestion
There are a lot of digital sensors out there. You can use a protocol like I2C
```

and SPI to read them. But analog sensors also exist! These sensors just output a
voltage level that's proportional to the magnitude they are sensing.

> The ADC peripheral can be use to convert that "analog" voltage level, say `1.25`

Reviewer: ```suggestion
The ADC peripheral can be used to convert that "analog" voltage level, say `1.25`
```

> Volts,into a "digital" number, say in the `[0, 65535]` range, that the processor

Reviewer: ```suggestion
Volts, into a "digital" number, say in the `[0, 65535]` range, that the processor
```

- SPI, abstracted within the [`embedded-hal` `spi` module] and implemented by the [`nrf52-hal`]
- I2S, currently not abstracted within the `embedded-hal` but implemented by the [`nrf52-hal`]
- Ethernet, there does exist a small TCP/IP stack named [`smoltcp`] which is implemented for some
  chips but the ones on the micro:bit don't feature an Ethernet peripheral
- USB, there is some experimental work on this, for example with the [`usb-device`] crate
- Bluetooth, there does exist an incomplete BLE stack named [`rubble`] which does support nrf chips.
- SMBUS, neither abstracted in `embedded-hal` nor implemented by the [`nrf52-hal`] at the moment.
- CAN, neither abstracted in `embedded-hal` nor implemented by the [`nrf52-hal`] at the moment
- IrDA, neither abstracted in `embedded-hal` nor implemented by the [`nrf52-hal`] at the moment

[`embedded-hal` `spi` module]: https://docs.rs/embedded-hal/0.2.6/embedded_hal/spi/index.html
[`smoltcp`]: https://github.com/smoltcp-rs/smoltcp
[`usb-device`]: https://github.com/mvirkkunen/usb-device
[`rubble`]: https://github.com/jonas-schievink/rubble

Different applications use different communication protocols. User facing

> applications usually have an USB connector because USB is an ubiquitous

Reviewer: ```suggestion
applications usually have a USB connector because USB is a ubiquitous
```

protocol in PCs and smartphones. Whereas inside cars you'll find plenty of CAN
"buses". Some digital sensors use SPI, others use I2C and others, SMBUS.

If you happen to be interested in developing abstractions in the `embedded-hal` or
implementations of peripherals in general, don't be shy to open an issue in the HAL
repositories. Alternatively you could also join the [Rust Embedded matrix channel]
and get into contact with most of the people who built the stuff from above.

## General Embedded-Relevant Topics

These topics cover items that are not specific to our device, or the hardware on
it. Instead, they discuss useful techniques that could be used on embedded
systems.

### Gyroscopes

As part of our Punch-o-meter exercise, we used the Accelerometer to measure

> changes in acceleration in three dimensions. But there are also other motion

Reviewer: ```suggestion
changes in acceleration in three dimensions. But there are other motion
```

> sensors such as gryoscopes, which allows us to measure changes in "spin" in three

Reviewer: ```suggestion
sensors such as gyroscopes, which allows us to measure changes in "spin" in three
```
